lifemonitor/auth/oauth2/server/models.py

```diff
@@ -0,0 +1,180 @@
+        )
+
+        client_metadata = {
+            "client_name": client_name,
+            "client_uri": client_uri,
+            "grant_types": grant_type,
+            "redirect_uris": redirect_uri,
+            "response_types": response_type,
+            "scope": scope,
+            "token_endpoint_auth_method": token_endpoint_auth_method
+        }
+        client.set_client_metadata(client_metadata)
+
+        if token_endpoint_auth_method == 'none':
+            client.client_secret = ''
+        else:
+            client.client_secret = gen_salt(48)
+
+        db.session.add(client)
+        db.session.commit()
+        return client
+
+
+class AuthorizationCode(db.Model, OAuth2AuthorizationCodeMixin):
+    id = db.Column(db.Integer, primary_key=True)
+    user_id = db.Column(
+        db.Integer, db.ForeignKey('user.id', ondelete='CASCADE')
+    )
+    user = db.relationship('User')
+
+
+class AuthorizationCodeGrant(grants.AuthorizationCodeGrant):
+
+    def create_authorization_code(self, client, grant_user, request):
+        # you can use other method to generate this code
+        code = generate_token(48)
+        item = AuthorizationCode(
+            code=code,
+            client_id=client.client_id,
+            redirect_uri=request.redirect_uri,
+            scope=request.scope,
+            user_id=grant_user.get_user_id(),
+        )
+        db.session.add(item)
+        db.session.commit()
+        return code
+
+    def parse_authorization_code(self, code, client):
+        item = AuthorizationCode.query.filter_by(
+            code=code, client_id=client.client_id).first()
+        if item and not item.is_expired():
+            return item
+
+    def delete_authorization_code(self, authorization_code):
+        db.session.delete(authorization_code)
+        db.session.commit()
+
+    def authenticate_user(self, authorization_code):
+        return User.query.get(authorization_code.user_id)
+
+
+class PasswordGrant(grants.ResourceOwnerPasswordCredentialsGrant):
+    TOKEN_ENDPOINT_AUTH_METHODS = [
+        'client_secret_basic', 'client_secret_post'
+    ]
+
+    def authenticate_user(self, username, password):
+        user = User.query.filter_by(username=username).first()
+        if user.check_password(password):
+            return user
+
+
+class ClientCredentialsGrant(grants.ClientCredentialsGrant):
+    TOKEN_ENDPOINT_AUTH_METHODS = [
+        'client_secret_basic', 'client_secret_post'
+    ]
+
+
+class RefreshTokenGrant(grants.RefreshTokenGrant):
+    TOKEN_ENDPOINT_AUTH_METHODS = [
+        'client_secret_basic', 'client_secret_post'
+    ]
+    INCLUDE_NEW_REFRESH_TOKEN = True
+
+    def authenticate_refresh_token(self, refresh_token):
+        item = Token.query.filter_by(refresh_token=refresh_token).first()
+        # define is_refresh_token_valid by yourself
+        # usually, you should check if refresh token is expired and revoked
+        if item and item.is_refresh_token_valid():
```

Comment:
need to return something if the if is False

Reply:
Fixed by https://github.com/crs4/life_monitor/pull/21/commits/a7381aa1f40c13f9811e173829f2da3245fef762